lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -128,6 +128,42 @@ void GenericTypeToArchetypeResolver::recordParamType(ParamDecl *decl, Type type)
         GenericEnv, type));
 }
 
+Type ProtocolRequirementTypeResolver::resolveGenericTypeParamType(
+    GenericTypeParamType *gp) {
+  assert(gp->getDepth() == 0 && gp->getIndex() == 0 &&
```

Comment:
Self with non-zero depth does show up in invalid code though... In the past explicit checks like this have always caused problems eventually. You should change this to compare against Proto->getSelfInterfaceType() eventually